fix: catalog cleanup

MatthewLemmond opened the pull request:
### Description
Addresses the following issues:
https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/issues/216
https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/issues/214
https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/issues/215
https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/issues/209
https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/issues/218

Also limits the prefix to address issues seen when using account-infrastructure-base in combination with landing zone

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [x] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

ocofaigh commented on variables.tf:
```diff
@@ -188,7 +188,7 @@ variable "api_creation" {
 variable "enforce_allowed_ip_addresses" {
   type        = bool
-  description = "If true IP address restriction will be enforced, If false, traffic originated outside specified allowed IP address set is monitored with audit events sent to SIEM and Activity Tracker. After running in monitored mode to test this variable, it should then explicitly be set to true to enforce IP allow listing, has no effect when `skip_iam_account_settings` is true."
+  description = "Whether the IP address restriction is enforced. If `false`, traffic originating outside of the specified allowed IP addresss is monitored with audit events sent to Activity Tracker. After running in monitoring mode to test the impact of the restriction, you must set to `true` to enforce the IP allowlist."
```
The default value is true here, so it should probably be re-worded to say if you want to test the impact of the restriction before enforcing it, set the value to false.

ocofaigh commented on solutions/account-infrastructure-base/variables.tf:
```diff
@@ -171,7 +175,7 @@ variable "api_creation" {
 variable "enforce_allowed_ip_addresses" {
   type        = bool
-  description = "Whether the IP address restriction is enforced. If `false`, traffic originating outside of the specified allowed IP addresss is monitored with audit events sent to QRadar SIEM and Activity Tracker. After running in monitoring mode to test the impact of the restriction, you must set to `true` to enforce the IP allowlist."
+  description = "Whether the IP address restriction is enforced. If `false`, traffic originating outside of the specified allowed IP addresss is monitored with audit events sent to Activity Tracker. After running in monitoring mode to test the impact of the restriction, you must set to `true` to enforce the IP allowlist."
```
The default value is true here, so it should probably be re-worded to say if you want to test the impact of the restriction before enforcing it, set the value to false.

ocofaigh commented on solutions/account-infrastructure-base/variables.tf:
```diff
@@ -19,6 +19,10 @@ variable "prefix" {
   default     = null
+  validation {
+    condition     = length(var.prefix) <= 13
+    error_message = "`prefix` length must be 13 characters or less."
```
where did you get 13 from? Suggest to leave a code comment explaining what module / service is restricting this

ocofaigh commented on solutions/account-infrastructure-base/variables.tf:
```diff
@@ -19,6 +19,10 @@ variable "prefix" {
   type        = string
   description = "An optional prefix to append to all resources created by this solution."
```
Suggest to mention that if this start with an upper case letter, the code is going to convert it to lower case for COS bucket naming if `provision_atracker_cos` is true

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+          "configuration": [
+            {
+              "key": "ibmcloud_api_key",
+              "type": "password"
```
no need to include the type here. It will automatically make it a password type due to the fact the terraform variable is marked as sensitive

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 3",
+                  "value": "NOT_SET"
```
`Option 1`, `Option 2`, `Option 3` makes no sense to the end user. How are they supposed to know what these options mean when they see them in the dropdown? 
Suggest this: `Restricted`, `Not restricted`, `Do not set`

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 6",
+                  "value": "NONE"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 3",
+                  "value": "NOT_SET"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+            },
+            {
+              "key": "user_mfa",
+              "type": "array"
```
no need to set the type

ocofaigh replied:
Can't we add a dropdown box here with the supported values?

MatthewLemmond replied:
I think this may be one instance where type may need to be specified because the underlying variable is a `set` and whenever I was testing the catalog was defaulting this to a string. I will remove the type to verify but if the catalog does default to string we will need to override here OR we can change the variable from a set to a list and find another way to enforce uniqueness

ocofaigh replied:
oh - well how about add the custom config with the dropdown for the supported values in the string format thats expected, and don't explicitly set any type. That should work. We can test it out too after you update this branch

MatthewLemmond replied:
So this variable is actually expecting a list of objects that looks like

```
[
  {
    iam_id = "IBMid-asdfqwerzxcv"
    mfa = LEVEL3
  },
  {
    iam_id = "IBMid-zxcvqwerasdf"
    mfa = LEVEL1
  }
]
```
setting per user MFA settings so a list (array for json) is the correct type, but whenever the type is not set in the catalog level it defaults to the string type.

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 2",
+                  "value": "cos-one-rate-plan"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 5",
+                  "value": "onerate_active"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 2",
+                  "value": "Accelerated"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+                },
+                {
+                  "displayname": "Option 3",
+                  "value": "direct"
```
same as above comment, these options mean nothing to end user

ocofaigh commented on ibm_catalog.json:
```diff
@@ -93,7 +93,366 @@
+            },
+            {
+              "key": "existing_audit_resource_group_name",
+              "custom_config": {
```
Your not going to be able to use the `resource_group` picker until catalog allow the default value of `null` to be auto selected. If you go with this now, the user is going to have to explicitly select null for all of those resource group picker vars you have, which is a terrible user experience. For now remove all custom config for all the resource group vars

ocofaigh reviewed (changes requested):
see comments

MatthewLemmond commented:
/run pipeline

MatthewLemmond commented:
/run pipeline

MatthewLemmond commented:
/run pipeline

MatthewLemmond commented:
/run pipeline

ocofaigh reviewed (changes requested):
Imported your code into the catalog, but no variables are showing. You must have an issue in your json. I suggest to backward engineer this (aka, edit a current version in the UI, make changes, and export the code so you can see the json)

![image](https://github.com/user-attachments/assets/67fb442e-1d66-4fa4-a45e-719654e682a5)

MatthewLemmond commented:
/run pipeline

MatthewLemmond commented:
/run pipeline

MatthewLemmond commented:
I've confirmed on import that the `user_mfa` field does not show up as an array, and terraform does not interpolate from a string to an array directly so I've added that override back into the json and added a default that matches the fscloud submodule of the COS module as we were previously relying on the default to be picked up from there but the catalog was giving errors since it wasn't tracing back to the other module to get its default.

ocofaigh commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.14.1 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-account-infrastructure-base/releases/tag/v1.14.1)
- `v1.14.1`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

vburckhardt commented:
I see a regression in latest version - prefix is now optional and set by default to "__NULL__" which causes validation failure.

```
 2024/11/27 11:04:52 Terraform plan | Error: Invalid function argument
 2024/11/27 11:04:52 Terraform plan | 
 2024/11/27 11:04:52 Terraform plan |   on variables.tf line 36, in variable "prefix":
 2024/11/27 11:04:52 Terraform plan |   36:     condition     = length(var.prefix) <= 13
 2024/11/27 11:04:52 Terraform plan |     ├────────────────
 2024/11/27 11:04:52 Terraform plan |     │ while calling length(value)
 2024/11/27 11:04:52 Terraform plan |     │ var.prefix is null
 2024/11/27 11:04:52 Terraform plan | 
 2024/11/27 11:04:52 Terraform plan | Invalid value for "value" parameter: argument must not be null.
 2024/11/27 11:04:53 [1m[31mTerraform PLAN error: Terraform PLAN errorexit status 1[39m[0m
```
